apps/payments/next/app/[locale]/[offeringId]/[interval]/new/page.tsx

```diff
@@ -118,6 +124,23 @@ export default async function New({
         fxaUid
       );
 
+      if (
+        searchParams.coupon &&
+        cart.eligibilityStatus === CartEligibilityStatus.UPGRADE
+      ) {
+        cart = await setupCartAction(
+          interval as SubplatInterval,
+          offeringId,
+          taxAddress,
+          undefined,
+          undefined,
+          fxaUid
+        );
+        const updatedSearchParams = { ...searchParams };
+        delete updatedSearchParams['coupon'];
+        searchParams = updatedSearchParams;
+      }
+
```

Comment:
issue (blocking): This would result in multiple carts being created unnecessarily.

suggestion: Move this check into `CartService.setupCart`.